coverage/artifacts/local.go

```diff
@@ -0,0 +1,99 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package artifacts
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+)
+
+//LocalArtifacts sub-type of Artifacts. Represent artifacts stored locally (
+// as oppose to artifacts stored in GCS bucket)
+type LocalArtifacts struct {
+	Artifacts
+}
+
+//NewLocalArtifacts constructs LocalArtifacts
+func NewLocalArtifacts(directory string, ProfileName string,
+	KeyProfileName string, CovStdoutName string) *LocalArtifacts {
+	return &LocalArtifacts{*New(
+		directory,
+		ProfileName,
+		KeyProfileName,
+		CovStdoutName)}
+}
+
+// ProfileReader create and returns a ProfileReader by opening the file stored in profile path
+func (arts *LocalArtifacts) ProfileReader() (io.ReadCloser, error) {
+	f, err := os.Open(arts.ProfilePath())
+	if err != nil {
+		logrus.Debugf("LocalArtifacts.ProfileReader(): os.Open(profilePath) error: %v", err)
+	}
+	return f, err
+}
+
+//ProfileName gets name of profile
+func (arts *LocalArtifacts) ProfileName() string {
+	return arts.profileName
+}
+
+// KeyProfileCreator creates a key profile file that will be used to hold a
+// filtered version of coverage profile that only stores the entries that
+// will be displayed by line coverage tool
+func (arts *LocalArtifacts) KeyProfileCreator() *os.File {
+	keyProfilePath := arts.KeyProfilePath()
+	keyProfileFile, err := os.Create(keyProfilePath)
+	logrus.Infof("os.Create(keyProfilePath)=%s", keyProfilePath)
+	if err != nil {
+		logrus.Fatalf("file(%s) creation error: %v", keyProfilePath, err)
+	}
+
+	return keyProfileFile
+}
+
+// ProduceProfileFile produce coverage profile (&its stdout) by running go test on target package
+// for periodic job, produce junit xml for testgrid in addition
+func (arts *LocalArtifacts) ProduceProfileFile(covTargetsStr string) error {
+	// creates artifacts directory
+	artsDirPath := arts.Directory
+	logrus.Infof("Making directory (MkdirAll): path=%s", artsDirPath)
+	if err := os.MkdirAll(artsDirPath, 0755); err != nil {
+		return fmt.Errorf("failed os.MkdirAll(path='%s', 0755); err='%v'", artsDirPath, err)
+	}
+	logrus.Infof("artifacts dir (path=%s) created successfully", artsDirPath)
+	covTargets := composeCmdArgs(covTargetsStr, arts.ProfilePath())
+	return runProfiling(covTargets, arts)
+}
+
+func composeCmdArgs(covTargetsStr, profileDestinationPath string) []string {
+	// generate the complete list of command line args for producing code coverage profile
+	var covTargets []string
+	for _, target := range strings.Split(covTargetsStr, " ") {
+		covTargets = append(covTargets, "./"+path.Join(target, "..."))
```

Comment:
What if `covTargetsStr` is `/absolute/path/with/ellipsis/...`?

Reply:
absolute path is not supported. Afaik, 'go test /absolute/path/with/ellipsis/...' will return an error with "matched no packages" - if you know how to make it work please let me know.

Reply:
we aren't handling this gracefully though. the user will get errors about some target pattern they didn't specify

Reply:
this goes back to this generally being incompatible with existing `go test ...` usage. you can also go test with a go import path instead of relative imports, go actually seems to prefer this.

Reply:
`go test k8s.io/test-infra/...` is perfectly cromulent

Reply:
Re-wrote the function